.emacs.d/config.org
#+title: Emacs configuration
#+author: Javier Pacheco
#+email: jpacheco@cock.li
#+description: My custom emacs configuration.
#+startup: content

* early-init.el
#+begin_src emacs-lisp :tangle ~/.emacs.d/early-init.el
;; Load theme first, so our eyes keep safe at startup...
(add-to-list 'load-path "~/.emacs.d/lisp/jp-themes")

(require 'jp-themes)

(defun jp/toggle-theme ()
  "Toggle between the `jp-eagle` and `jp-autumn` themes."
  (interactive)
  (let ((current-theme (car custom-enabled-themes)))
    (if (eq current-theme 'jp-eagle)
        (progn
          ;; (disable-theme 'jp-eagle)
          (jp-themes-select 'miasma)
          (message "Dark theme loaded."))
      (progn
        (jp-themes-select 'jp-eagle)
        (message "Light theme loaded.")))))

(mapc #'disable-theme custom-enabled-themes)
;; Load the theme of choice:
;; (load-theme 'jp-darkvenom :no-confirm)
;; (load-theme 'jp-gruvby :no-confirm)
;; (load-theme 'jp-dream :no-confirm)
(load-theme 'miasma :no-confirm)
;; (load-theme 'jp-elea-dark :no-confirm)
;; (load-theme 'jp-owl :no-confirm)
(global-set-key (kbd "<f10>") 'jp/toggle-theme)

;; Disable GUI when foundit
(when (fboundp 'menu-bar-mode)
  (menu-bar-mode -1))
(when (fboundp 'tool-bar-mode)
  (tool-bar-mode -1))
(when (fboundp 'scroll-bar-mode)
  (scroll-bar-mode -1))
(when (fboundp 'horizontal-scroll-bar-mode)
  (horizontal-scroll-bar-mode -1))

;; Initial message
(setq initial-scratch-message
	  ";;; -*- Calling emacs an editor is like calling the Earth a hunk of dirt.  -*- lexical-binding: t; -*-
;; --
;; It’s difficult to be rigorous about whether a machine really ’knows’, ’thinks’, etc.,
;; because we’re hard put to define these things. We understand human mental processes
;; only slightly better than a fish understands swimming.
;; --
;; <Jhon McCarthy>\n\n
;;; Code:\n")

;; (setq initial-scratch-message
;; 	  ";;; -*- Calling emacs an editor is like calling the Earth a hunk of dirt.  -*- lexical-binding: t; -*-
;; ;; --
;; ;; There are two ways to construct a software design:
;; ;; One is to make it so simple that it is obvious that there are no deficiencies,
;; ;; and the other is to make it so complicated that there are no obvious deficiencies.
;; ;; --
;; ;; <C. A. R. Hoare>\n\n
;; ;;; Code:\n")

;; Initialize package sources
(add-to-list 'load-path "~/.emacs.d/lisp/")
;; (require 'nano-splash)	;; Splash screen
;; (require 'buffer-move)   	;; Buffer-move for better window management
(require 'utilities)		;; Custom scripts
;; (require 'term-toggle)	;; toggle-term

#+end_src
* init.el
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el
(setq byte-compile-warnings '(cl-functions))
(defun my-generate-tab-stops (&optional width max)
  "Return a sequence suitable for `tab-stop-list'."
  (let* ((max-column (or max 200))
		 (tab-width (or width tab-width))
		 (count (/ max-column tab-width)))
	(number-sequence tab-width (* tab-width count) tab-width)))

(setq blink-cursor-mode t)
(setq ring-bell-function 'ignore)
(setq tab-stop-list (my-generate-tab-stops))
(put 'outline-forward-same-level 'disabled t)
(setq-default indent-tabs-mode t)
(setq-default tab-width 4) ; Assuming you want your tabs to be four spaces wide
(defvaralias 'c-basic-offset 'tab-width)

(setq visible-bell t)

(setq-default truncate-lines t)
(setq-default fill-column 80)
(setq line-move-visual t) ;; C-p, C-n, etc uses visual lines
(setq-default display-line-numbers-width 3)

;; Don't generate backups or lockfiles.
(setq create-lockfiles nil
      make-backup-files nil
      ;; But in case the user does enable it, some sensible defaults:
      version-control t     ; number each backup file
      backup-by-copying t   ; instead of renaming current file (clobbers links)
      delete-old-versions t ; clean up after itself
      kept-old-versions 5
      kept-new-versions 5)

(setq dired-kill-when-opening-new-dired-buffer t)
(setq package-install-upgrade-built-in t)

;; The default is 800 kilobytes.  Measured in bytes.
(setq gc-cons-threshold (* 50 1000 1000))

(defun jp/display-startup-time ()
  (message "Emacs loaded in %s with %d garbage collections."
           (format "%.2f seconds"
                   (float-time
                    (time-subtract after-init-time before-init-time)))
           gcs-done))

(add-hook 'emacs-startup-hook #'jp/display-startup-time)
(add-hook 'emacs-startup-hook 'blink-cursor-mode)

;; (org-babel-load-file
;;  (expand-file-name
;;   "config.org"
;;   user-emacs-directory))

(setq custom-file "~/.emacs.d/jp-config.el")
(setq org-config-file "~/.emacs.d/config.org")

(if (file-exists-p custom-file)
    ;; If the custom file exists, load it directly
    (load custom-file)
  ;; If the custom file doesn't exist, tangle it from the Org file and then load it
  (when (file-exists-p org-config-file)
    (require 'org)
    (org-babel-tangle-file org-config-file custom-file)
    (load custom-file)))

;; Fonts settings.

(set-face-attribute 'default nil
                    :family "Fira Code"
                    :height 130
                    :weight 'regular)

;; Set italic font
(set-face-attribute 'italic nil
                    :family "Fira Code"
                    :height 130
                    :slant 'italic)

;; Set bold font
(set-face-attribute 'bold nil
                    :family "Fira Code"
                    :height 130
                    :weight 'bold)

;; ;; Set bold-italic font
;; (set-face-attribute 'bold-italic nil
;;                     :family "JetBrains Mono"
;;                     :slant 'italic)

;; Set font for comments to be italic
(set-face-attribute 'font-lock-comment-face nil
                    :family "Fira Code"
                    :slant 'italic)

;; Optionally, also set italic for doc comments
(set-face-attribute 'font-lock-doc-face nil
                    ;; :family "IBM Plex Mono"
                    :family "Fira Code"
                    :slant 'italic)

;; Set monospaced font for code and programming modes
(set-face-attribute 'org-block nil
                    ;; :family "IBM Plex Mono"
                    :family "Fira Code"
                    :height 120)

;; Optionally, set the code block font (Org-mode source blocks, markdown, etc.)
(set-face-attribute 'org-verbatim nil
                    ;; :family "IBM Plex Mono"
                    :family "Fira Code"
					:slant 'italic
                    :height 130)


(custom-set-faces
 ;; custom-set-faces was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(evil-goggles-change-face ((t (:inherit diff-removed))))
 '(evil-goggles-delete-face ((t (:inherit diff-removed))))
 '(evil-goggles-paste-face ((t (:inherit diff-added))))
 '(evil-goggles-undo-redo-add-face ((t (:inherit diff-added))))
 '(evil-goggles-undo-redo-change-face ((t (:inherit diff-changed))))
 '(evil-goggles-undo-redo-remove-face ((t (:inherit diff-removed))))
 '(evil-goggles-yank-face ((t (:inherit diff-changed))))
 '(org-checkbox ((t (:box (:style released-button)))))
 '(org-checkbox-statistics-done ((t (:inherit org-todo))))

 ;; Enable org-crypt and epa-file
 (require 'org-crypt)
 (require 'epa-file)

 ;; Enable epa-file for encryption/decryption of files
 (epa-file-enable)

 (setq epa-file-encrypt-to '("jpacheco@cock.li"))  ; Replace with your GPG key email

 ;; Configure org-crypt
 (setq org-crypt-tag-matcher "crypt")                    ; Tag used to encrypt entries
 (setq org-crypt-key "jpacheco@cock.li")          ; Replace with your GPG key email
 (setq org-tags-exclude-from-inheritance '("crypt"))     ; Prevent inheritance of "crypt" tag
 (setq epa-pinentry-mode 'loopback) ;; Ensures passphrase prompt in minibuffer

 ;; Automatically encrypt entries tagged with "crypt" before saving
 (add-hook 'before-save-hook 'org-crypt-use-before-save-magic)

 ;; Define keybindings for manual encryption/decryption
 (define-key org-mode-map (kbd "C-c C-x e") 'org-encrypt-entries)
 (define-key org-mode-map (kbd "C-c C-x d") 'org-decrypt-entries)
 )
#+end_src
* Package Manager and some extra files.
#+BEGIN_SRC emacs-lisp :tangle jp-config.el
;; ;; Initialize package sources
;; (add-to-list 'load-path "~/.emacs.d/lisp/")
;; (require 'buffer-move)   	;; Buffer-move for better window management
;; (require 'utilities)		;; Custom scripts
;; (require 'term-toggle)	;; toggle-term
;; (require 'nano-splash)	;; Splash screen

(require 'package)
(setq package-archives '(("melpa" . "https://melpa.org/packages/")
                        ;; ("org" . "https://orgmode.org/elpa/")
                        ("gnu" . "https://elpa.gnu.org/packages/")
                        ("elpa" . "https://elpa.gnu.org/packages/")))
(package-initialize)
(unless package-archive-contents
  (package-refresh-contents))
;; Initialize use-package on non-Linux platforms
(unless (package-installed-p 'use-package)
(package-install 'use-package))
(require 'use-package)
(setq use-package-always-ensure t)

(use-package auto-package-update
  :custom
  (auto-package-update-interval 7)
  (auto-package-update-prompt-before-update t)
  (auto-package-update-hide-results t)
  :config
  (auto-package-update-maybe)
  (auto-package-update-at-time "09:00"))

(use-package no-littering)

;; no-littering doesn't set this by default so we must place
;; auto save files in the same path as it uses for sessions
(setq auto-save-file-name-transforms
      `((".*" ,(no-littering-expand-var-file-name "auto-save/") t)))

(use-package async
  :config (setq async-bytecomp-package-mode 1))
#+END_SRC
* Keycast.
#+BEGIN_SRC emacs-lisp :tangle jp-config.el
(use-package keycast
  :hook (after-init . keycast-mode)
  :config
  (setopt keycast-mode-line-format
          "%k%r")
  (setopt keycast-substitute-alist
          '((keycast-log-erase-buffer nil nil)
            (transient-update         nil nil)
            (self-insert-command      nil nil)
            (mwheel-scroll nil nil)))
  (define-minor-mode keycast-mode
	"Show current command and its key binding in the mode line (fix for use with doom-modeline)."
	:global t
	(if keycast-mode
		(add-hook 'pre-command-hook 'keycast--update t)
      (remove-hook 'pre-command-hook 'keycast--update)))

  (add-to-list 'global-mode-string '("" keycast-mode-line)))
#+END_SRC
* Custom modeline.
#+BEGIN_SRC emacs-lisp :tangle jp-config.el
;; Load modeline
;; (require 'custom-modeline)
;; (custom-modeline-mode)
;; (setq custom-modeline-segments
;;       `((custom-modeline-segment-modified	;; icon if file is modified 
;; 		 custom-modeline-segment-buffer-name)	;; buffer name
;; 		(custom-modeline-segment-major-mode 	;; Major mode
;; 		 custom-modeline-segment-vc 			;; vc status
;; 		 custom-modeline-segment-misc-info 	;; Clock and frame name
;; 		 custom-modeline-segment-process)))

(use-package doom-modeline
  :ensure t
  :hook (after-init . doom-modeline-mode))

;; (with-eval-after-load 'doom-modeline
;;   (doom-modeline-def-segment lsp
;;     "Displays LSP server status."
;;     (when (and (bound-and-true-p lsp-mode) (lsp-workspaces))
;;       (concat
;;        (propertize (doom-modeline-spc) 'face (if (doom-modeline--active) 'mode-line 'mode-line-inactive))
;;        (propertize (nerd-icons-mdicon "nf-md-repeat")
;;                    'face `(:family ,(nerd-icons-mdicon-family) :inherit))
;;                    ;; 'display '(raise -0.1))
;;        (propertize (doom-modeline-spc) 'face (if (doom-modeline--active) 'mode-line 'mode-line-inactive))))))

(use-package hide-mode-line
  :ensure t
  :defer t)

(setq doom-modeline-icon nil)
(setq doom-modeline-enable-word-count nil)
(setq doom-modeline-time-analogue-clock nil)
(setq doom-modeline-position-line-format nil)
(setq doom-modeline-buffer-encoding nil)
(setq doom-modeline-percent-position '(-3 "%p"))
(setq display-time-default-load-average nil)
(setq display-time-load-average nil)
(display-time-mode -1) ;; displays current time
;; Default custom modeline
(setq-default mode-line-format (delq 'mode-line-modes mode-line-format))
(setq doom-modeline-modal-icon nil)
(setq evil-normal-state-tag   (propertize "[Normal]" )
      evil-emacs-state-tag    (propertize "[Emacs]" )
      evil-insert-state-tag   (propertize "[Insert]" )
      evil-motion-state-tag   (propertize "[Motion]" )
      evil-visual-state-tag   (propertize "[Visual]" )
      evil-operator-state-tag (propertize "[Operator]" ))

(setq evil-insert-state-message nil)
(setq evil-visual-state-message nil)
(setq evil-replace-state-message nil)
(setq evil-motion-state-message nil)
(setq hl-line-sticky-flag t)
;; (global-hl-line-mode 1)

;; (use-package nyan-mode
;;   :ensure t
;;   :init (nyan-mode))

;; (defun my/toggle-nyan-mode ()
;;   "Enable or disable `nyan-mode` depending on whether the buffer needs scrolling."
;;   (if (>= (line-number-at-pos (point-max)) (window-body-height))
;;       (nyan-mode 1)  ;; Enable nyan-mode if the buffer has more lines than the window height
;;     (nyan-mode -1))) ;; Disable nyan-mode otherwise

;; (defun my/setup-nyan-mode-toggle ()
;;   "Set up automatic toggling of `nyan-mode`."
;;   (add-hook 'window-size-change-functions #'my/toggle-nyan-mode)
;;   (add-hook 'post-command-hook #'my/toggle-nyan-mode))

;; (my/setup-nyan-mode-toggle)

(use-package diminish
  :ensure t)
#+END_SRC

#+RESULTS:

* Icons/Eyecandy Stuff.
#+begin_src elisp :tangle jp-config.el
(use-package all-the-icons-completion
  :ensure t
  :config
  (all-the-icons-completion-mode))

(use-package nerd-icons
:ensure t)

(use-package nerd-icons-completion
  :after marginalia
  :config
  (nerd-icons-completion-mode)
  (add-hook 'marginalia-mode-hook #'nerd-icons-completion-marginalia-setup))

 ;; (use-package pretty-mode
 ;;   :defer t 
 ;;   :ensure t
 ;;   :hook
 ;;   (prog-mode . pretty-mode))
#+END_SRC
* Evil-mode.
** Evil.
#+BEGIN_SRC emacs-lisp :tangle jp-config.el
(use-package evil
  :init      ;; tweak evil's configuration before loading it
  (setq evil-want-integration t  ;; This is optional since it's already set to t by default.
  evil-want-keybinding nil
  evil-vsplit-window-right t
  evil-split-window-below t
  evil-undo-system 'undo-redo)  ;; Adds vim-like C-r redo functionality
  :config
  (evil-mode))

(defun my/evil-open-at-point ()
  "Open link at point in org-mode or run `dashboard-return` in dashboard-mode. Do nothing elsewhere."
  (interactive)
  (cond
   ((eq major-mode 'org-mode)
    (org-open-at-point))

   ((eq major-mode 'dashboard-mode)
    (when (fboundp 'dashboard-return)
      (dashboard-return)))

   (t nil))) ;; do nothing

(define-key evil-normal-state-map (kbd "RET") #'my/evil-open-at-point)
#+END_SRC

#+RESULTS:
: my/evil-org-ret-or-ignore

** Evil-collection.
#+BEGIN_SRC emacs-lisp :tangle jp-config.el
(use-package evil-collection
  :after evil
  :config
      (add-to-list 'evil-collection-mode-list 'help) ;; evilify help mode
      (evil-collection-init))
#+END_SRC
** Evil-commenter.
#+begin_src emacs-lisp :tangle jp-config.el
(use-package evil-nerd-commenter
  :ensure t
  :after evil)
#+end_src
** Evil-numbers.
#+begin_src emacs-lisp :tangle jp-config.el
(use-package evil-numbers
  :ensure t
  :after evil)
(define-key evil-normal-state-map (kbd "C-a") 'evil-numbers/inc-at-pt)
(define-key evil-normal-state-map (kbd "C-s") 'evil-numbers/dec-at-pt)
#+end_src
** Evil-surround.
#+BEGIN_SRC emacs-lisp :tangle jp-config.el
(use-package evil-surround
  :ensure t
  :after evil
  :config
  (global-evil-surround-mode 1))

;; Using RETURN to follow links in Org/Evil 
(with-eval-after-load 'evil-maps
  (define-key evil-motion-state-map (kbd "SPC") nil)
  (define-key evil-motion-state-map (kbd "RET") nil)
  (define-key evil-motion-state-map (kbd "TAB") nil))
#+END_SRC
** Evil-Goggles.
#+BEGIN_SRC emacs-lisp :tangle jp-config.el
(use-package evil-goggles
  :ensure t
  :after evil
  :config
  (evil-goggles-mode)
  (evil-goggles-use-diff-faces))
#+END_SRC
** Evil-owl.
#+BEGIN_SRC emacs-lisp :tangle jp-config.el
(use-package evil-owl
  :config
  (setq evil-owl-display-method 'posframe
        evil-owl-extra-posframe-args '(:width 50 :height 20)
        evil-owl-max-string-length 50)
  (evil-owl-mode))
#+END_SRC
** Evil-terminal-changer
#+begin_src emacs-lisp :tangle jp-config.el
(use-package evil-terminal-cursor-changer
  :ensure t)
(setq evil-motion-state-cursor 'box)  ; █
(setq evil-visual-state-cursor 'box)  ; █
(setq evil-normal-state-cursor 'box)  ; █
(setq evil-insert-state-cursor 'hbar)  ; ⎸
(setq evil-emacs-state-cursor  'hbar) ; _
#+end_src
* keychord.
#+BEGIN_SRC emacs-lisp :tangle jp-config.el
(use-package key-chord
  :ensure t
  :after evil
  :config
  (key-chord-mode 1)
  ;; Use "jk" to exit insert mode (similar to "jj")
  (key-chord-define evil-insert-state-map "jk" 'evil-normal-state)
  (key-chord-define evil-insert-state-map "jj" 'evil-normal-state)
  ;; Move to the beginning or end of the line with "hh" or "ll"
  (key-chord-define evil-insert-state-map "hh" 'move-beginning-of-line)
  (key-chord-define evil-insert-state-map  "l;" 'move-end-of-line)
  (key-chord-define evil-insert-state-map  "aa" 'move-end-of-line)
  (key-chord-define evil-normal-state-map  "sc" 'evil-avy-goto-char-2)
  (setq key-chord-two-keys-delay 0.5))
#+END_SRC
* Corfu.
#+BEGIN_SRC emacs-lisp :tangle no
(use-package corfu
  ;; TAB-and-Go customizations
  :custom
  (corfu-cycle t)                 ; Allows cycling through candidates
  (corfu-auto nil)                  ; Enable auto completion
  (corfu-auto-prefix 1)
  (corfu-auto-delay 0.2)
  (corfu-popupinfo-delay '(0.5 . 0.5))
  (corfu-preview-current 'valid) ; insert previewed candidate
  (corfu-preselect 'prompt)
  (corfu-on-exact-match nil)      ; Don't auto expand tempel snippets
  (corfu-min-width 40)
  (corfu-max-width corfu-min-width)     ; Always have the same width
  (corfu-count 14)
  (corfu-scroll-margin 4)
  (corfu-right-margin-width 1.0) ; Give some margin to align with nerd-icons left margin.
  (corfu-left-margin-width 1.0)
  (corfu-quit-at-boundary nil)
  (corfu-separator ?\s)            ; Use space
  (corfu-quit-no-match 'separator) ; Don't quit if there is `corfu-separator' inserted
  (corfu-preview-current 'insert)  ; Preview first candidate. Insert on input if only one
  (corfu-preselect-first t)        ; Preselect first candidate?

  ;; Use TAB for cycling, default is `corfu-complete'.
  :bind
  (:map corfu-map
		("TAB" . corfu-next)
		([tab] . corfu-next)
		("S-TAB" . corfu-previous)
		([backtab] . corfu-previous))

  :init
  ;; (global-corfu-mode)
  (corfu-history-mode)
  (corfu-popupinfo-mode) ; Popup completion info
  :config
  (add-hook 'eshell-mode-hook
			(lambda () (setq-local corfu-quit-at-boundary t
							  corfu-quit-no-match t
							  corfu-auto nil) (corfu-mode)) nil t)

  :hook ((prog-mode . corfu-mode)
		 (text-mode . corfu-mode)))


;; (use-package corfu-doc
;;   :after corfu
;;   :config
;;   (define-key corfu-map (kbd "M-p") #'corfu-doc-scroll-down) ;; corfu-next
;;   (define-key corfu-map (kbd "M-n") #'corfu-doc-scroll-up)  ;; corfu-previous
;;   (setq corfu-doc-display-within-parent-frame nil)
;;   (add-hook 'corfu-mode-hook #'corfu-doc-mode))
;; (define-key corfu-map (kbd "M-d") #'corfu-doc-toggle)

;; Icons
(use-package kind-icon
  :ensure t
  :after corfu
  :custom
  (kind-icon-use-icons t)
  (kind-icon-default-face 'corfu-default) ; Have background color be the same as `corfu' face background
  (kind-icon-blend-background nil)  ; Use midpoint color between foreground and background colors ("blended")?
  (kind-icon-blend-frac 0.08)
  :config
  (add-to-list 'corfu-margin-formatters #'kind-icon-margin-formatter))

(setq kind-icon-mapping
	  '((array          "a"   :icon "symbol-array"       :face font-lock-type-face              :collection "vscode")
		(boolean        "b"   :icon "symbol-boolean"     :face font-lock-builtin-face           :collection "vscode")
		(color          "#"   :icon "symbol-color"       :face success                          :collection "vscode")
		(command        "cm"  :icon "chevron-right"      :face default                          :collection "vscode")
		(constant       "co"  :icon "symbol-constant"    :face font-lock-constant-face          :collection "vscode")
		(class          "c"   :icon "symbol-class"       :face font-lock-type-face              :collection "vscode")
		(constructor    "cn"  :icon "symbol-method"      :face font-lock-function-name-face     :collection "vscode")
		(enum           "e"   :icon "symbol-enum"        :face font-lock-builtin-face           :collection "vscode")
		(enummember     "em"  :icon "symbol-enum-member" :face font-lock-builtin-face           :collection "vscode")
		(enum-member    "em"  :icon "symbol-enum-member" :face font-lock-builtin-face           :collection "vscode")
		(event          "ev"  :icon "symbol-event"       :face font-lock-warning-face           :collection "vscode")
		(field          "fd"  :icon "symbol-field"       :face font-lock-variable-name-face     :collection "vscode")
		(file           "f"   :icon "symbol-file"        :face font-lock-string-face            :collection "vscode")
		(folder         "d"   :icon "folder"             :face font-lock-doc-face               :collection "vscode")
		(function       "f"   :icon "symbol-method"      :face font-lock-function-name-face     :collection "vscode")
		(interface      "if"  :icon "symbol-interface"   :face font-lock-type-face              :collection "vscode")
		(keyword        "kw"  :icon "symbol-keyword"     :face font-lock-keyword-face           :collection "vscode")
		(macro          "mc"  :icon "lambda"             :face font-lock-keyword-face)
		(magic          "ma"  :icon "lightbulb-autofix"  :face font-lock-builtin-face           :collection "vscode")
		(method         "m"   :icon "symbol-method"      :face font-lock-function-name-face     :collection "vscode")
		(module         "{"   :icon "file-code-outline"  :face font-lock-preprocessor-face)
		(numeric        "nu"  :icon "symbol-numeric"     :face font-lock-builtin-face           :collection "vscode")
		(operator       "op"  :icon "symbol-operator"    :face font-lock-comment-delimiter-face :collection "vscode")
		(param          "pa"  :icon "gear"               :face default                          :collection "vscode")
		(property       "pr"  :icon "symbol-property"    :face font-lock-variable-name-face     :collection "vscode")
		(reference      "rf"  :icon "library"            :face font-lock-variable-name-face     :collection "vscode")
		(snippet        "S"   :icon "symbol-snippet"     :face font-lock-string-face            :collection "vscode")
		(string         "s"   :icon "symbol-string"      :face font-lock-string-face            :collection "vscode")
		(struct         "%"   :icon "symbol-structure"   :face font-lock-variable-name-face     :collection "vscode")
		(text           "tx"  :icon "symbol-key"         :face font-lock-doc-face               :collection "vscode")
		(typeparameter  "tp"  :icon "symbol-parameter"   :face font-lock-type-face              :collection "vscode")
		(type-parameter "tp"  :icon "symbol-parameter"   :face font-lock-type-face              :collection "vscode")
		(unit           "u"   :icon "symbol-ruler"       :face font-lock-constant-face          :collection "vscode")
		(value          "v"   :icon "symbol-enum"        :face font-lock-builtin-face           :collection "vscode")
		(variable       "va"  :icon "symbol-variable"    :face font-lock-variable-name-face     :collection "vscode")
		(t              "."   :icon "question"           :face font-lock-warning-face           :collection "vscode")))

;; yasnippets integration with corfu.
(use-package yasnippet-capf
  :after cape)

(use-package cape
  :bind (("C-c p p" . completion-at-point) ;; capf
         ("C-c p t" . complete-tag)        ;; etags
         ("C-c p d" . cape-dabbrev)        ;; or dabbrev-completion
         ("C-c p h" . cape-history)
         ("C-c p f" . cape-file)
         ("C-c p k" . cape-keyword)
         ("C-c p s" . cape-elisp-symbol)
         ("C-c p e" . cape-elisp-block)
         ("C-c p a" . cape-abbrev)
         ("C-c p l" . cape-line)
         ("C-c p w" . cape-dict)
         ("C-c p :" . cape-emoji)
         ("C-c p \\" . cape-tex)
         ("C-c p _" . cape-tex)
         ("C-c p ^" . cape-tex)
         ("C-c p &" . cape-sgml)
         ("C-c p r" . cape-rfc1345))
  :init
  (add-hook 'completion-at-point-functions #'cape-dabbrev)
  (add-hook 'completion-at-point-functions #'cape-file)
  (add-hook 'completion-at-point-functions #'cape-elisp-block))

;; Function to prioritize yasnippet-capf in completion-at-point-functions
(defun my-prioritize-yasnippet-capf ()
  "Ensure `yasnippet-capf` is the first in `completion-at-point-functions`."
  (let ((yas-capf (car (remove 'yasnippet-capf completion-at-point-functions))))
    (setq completion-at-point-functions
          (cons 'yasnippet-capf (remove 'yasnippet-capf completion-at-point-functions)))))

;; Hook the function to major modes where you want this behavior
(add-hook 'prog-mode-hook #'my-prioritize-yasnippet-capf)
(add-hook 'text-mode-hook #'my-prioritize-yasnippet-capf)
#+END_SRC
* Company.
#+begin_src emacs-lisp :tangle jp-config.el
(use-package company
  :ensure t
  :hook (prog-mode . company-mode)   ;; Only activate in prog-mode
  :config
  (setq company-dabbrev-downcase 0)
  (setq company-idle-delay 0.1)
  (setq company-minimum-prefix-length 1)
  (setq company-icon-size 25)
  (setq company-tooltip-align-annotations t)
  
  ;; Use TAB to select completion
  (define-key company-active-map (kbd "<tab>") 'company-complete-selection)
  (define-key company-active-map (kbd "TAB") 'company-complete-selection)
  
  ;; Use C-j and C-k for navigation
  (define-key company-active-map (kbd "C-j") 'company-select-next)
  (define-key company-active-map (kbd "C-k") 'company-select-previous)
  
  ;; Allow TAB to complete even when the selection isn't explicitly highlighted
  (setq company-selection-wrap-around t))

(setq company-tooltip-align-annotations t ; aligns annotation to the right
      company-tooltip-limit 12 ; bigger popup window
      company-idle-delay 0.5   ; decrease delay before autocompletion popup shows
      company-echo-delay (if (display-graphic-p) nil 0)
      company-minimum-prefix-length 2
      company-require-match nil
      company-dabbrev-ignore-case nil
      company-dabbrev-downcase nil
      company-dabbrev-minimum-length 10 ; Use buffer completion at 10 charcters.
      company-global-modes '(not erc-mode message-mode help-mode
                                 gud-mode eshell-mode shell-mode)
      ;;company-backends '((company-capf :with company-yasnippet)
      ;;                   (company-dabbrev-code company-keywords company-files)
      ;;                   company-dabbrev))

      ;; This puts snippet completion first.
      company-backends '((:separate company-yasnippet company-files company-bbdb company-semantic company-cmake company-capf :with company-tempo company-clang

.config/ags/options.ts
import { opt, mkOptions } from "lib/option"
import { distro } from "lib/variables"
import { icon } from "lib/utils"
import icons from "lib/icons"

const options = mkOptions(OPTIONS, {
    autotheme: opt(true),

    wallpaper: {
        resolution: opt<import("service/wallpaper").Resolution>(1920),
        market: opt<import("service/wallpaper").Market>("random"),
    },

    theme: {
        dark: {
            primary: {
                bg: opt("#458588"),
                fg: opt("#282828"),
            },
            error: {
                bg: opt("#e55f86"),
                fg: opt("#141414"),
            },
            bg: opt("#282828"),
            fg: opt("#eeeeee"),
            widget: opt("#eeeeee"),
            border: opt("#458588"),
        },
        light: {
            primary: {
                bg: opt("#426ede"),
                fg: opt("#eeeeee"),
            },
            error: {
                bg: opt("#b13558"),
                fg: opt("#eeeeee"),
            },
            bg: opt("#fffffa"),
            fg: opt("#121318"),
            widget: opt("#121318"),
            border: opt("#458588"),
        },

        blur: opt(0),
        scheme: opt<"dark" | "light">("dark"),
        widget: { opacity: opt(94) },
        border: {
            width: opt(0),
            opacity: opt(96),
        },

        shadows: opt(false),
        padding: opt(5),
        spacing: opt(5),
        radius: opt(5),
    },

    transition: opt(200),

    font: {
        size: opt(13),
        name: opt("JetBrains Mono"),
    },

    bar: {
        flatButtons: opt(false),
        position: opt<"top" | "bottom">("top"),
        corners: opt(0),
        transparent: opt(false),
        layout: {
            start: opt<Array<import("widget/bar/Bar").BarWidget>>([
                "launcher",
                "workspaces",
                // "taskbar",
                "expander",
            ]),
          center: opt<Array<import("widget/bar/Bar").BarWidget>>([
                "date",
            ]),
            end: opt<Array<import("widget/bar/Bar").BarWidget>>([
                // "media",
              "expander",
              // "systray",
               // "colorpicker",
              "screenrecord",
               "system",
              "messages",
              "battery",
              "powermenu",
            ]),
        },
        launcher: {
            icon: {
                colored: opt(true),
                icon: opt(icon(distro.logo, icons.ui.search)),
            },
            label: {
                colored: opt(false),
                label: opt(""),
            },
            action: opt(() => App.toggleWindow("launcher")),
        },
        date: {
            format: opt("%r - %A %e"),
            action: opt(() => App.toggleWindow("datemenu")),
        },
        battery: {
            bar: opt<"hidden" | "regular" | "whole">("regular"),
            charging: opt("#b7d085"),
            percentage: opt(true),
            blocks: opt(7),
            width: opt(50),
            low: opt(30),
        },
        workspaces: {
            workspaces: opt(10),
        },
        taskbar: {
            iconSize: opt(0),
            monochrome: opt(true),
            exclusive: opt(false),
        },
        messages: {
            action: opt(() => App.toggleWindow("datemenu")),
        },
        systray: {
            ignore: opt([
                "KDE Connect Indicator",
                "spotify-client",
            ]),
        },
        media: {
            monochrome: opt(true),
            preferred: opt("spotify"),
            direction: opt<"left" | "right">("right"),
            format: opt("{artists} - {title}"),
            length: opt(40),
        },
        powermenu: {
            monochrome: opt(false),
            action: opt(() => App.toggleWindow("powermenu")),
        },
    },

    launcher: {
        width: opt(0),
        margin: opt(80),
        nix: {
            pkgs: opt("nixpkgs/nixos-unstable"),
            max: opt(8),
        },
        sh: {
            max: opt(16),
        },
        apps: {
            iconSize: opt(50),
            max: opt(6),
            favorites: opt([
                [
                    "footclient",
                    "Emacs",
                    "firefox",
                ],
            ]),
        },
    },

    overview: {
        scale: opt(9),
        workspaces: opt(10),
        monochromeIcon: opt(true),
    },

    powermenu: {
        sleep: opt("systemctl suspend"),
        reboot: opt("systemctl reboot"),
        logout: opt("pkill Hyprland"),
        shutdown: opt("shutdown now"),
        layout: opt<"line" | "box">("line"),
        labels: opt(true),
    },

    quicksettings: {
        avatar: {
            image: opt(`/home/javier/.config/hypr/me.jpg`),
            size: opt(70),
        },
        width: opt(380),
        position: opt<"left" | "center" | "right">("right"),
        networkSettings: opt("gtk-launch gnome-control-center"),
        media: {
            monochromeIcon: opt(true),
            coverSize: opt(100),
        },
    },

    datemenu: {
        position: opt<"left" | "center" | "right">("center"),
        weather: {
            interval: opt(60_000),
            unit: opt<"metric" | "imperial" | "standard">("metric"),
            key: opt<string>(
                JSON.parse(Utils.readFile(`${App.configDir}/.weather`) || "{}")?.key || "",
            ),
            cities: opt<Array<number>>(
                JSON.parse(Utils.readFile(`${App.configDir}/.weather`) || "{}")?.cities || [],
            ),
        },
    },

    osd: {
        progress: {
            vertical: opt(true),
            pack: {
                h: opt<"start" | "center" | "end">("end"),
                v: opt<"start" | "center" | "end">("center"),
            },
        },
        microphone: {
            pack: {
                h: opt<"start" | "center" | "end">("center"),
                v: opt<"start" | "center" | "end">("end"),
            },
        },
    },

    notifications: {
        position: opt<Array<"top" | "bottom" | "left" | "right">>(["top", "right"]),
        blacklist: opt(["Spotify"]),
        width: opt(440),
    },

    hyprland: {
        gaps: opt(2.4),
        inactiveBorder: opt("#282828"),
        gapsWhenOnly: opt(true),
    },
})

globalThis["options"] = options
export default options
